packages/amplify_api/lib/src/graphql/graphql_response_decoder.dart

```diff
@@ -0,0 +1,37 @@
+import 'dart:convert';
+
+import 'package:amplify_api/amplify_api.dart';
+
+class GraphQLResponseDecoder {
+  // Singleton methods/properties
+  // usage: GraphQLResponseDecoder.instance;
+  GraphQLResponseDecoder._();
+
+  static final GraphQLResponseDecoder _instance = GraphQLResponseDecoder._();
+
+  static GraphQLResponseDecoder get instance => _instance;
+
+  GraphQLResponse<T> decode<T>(
+      {required GraphQLRequest request,
+      required String data,
+      required List<GraphQLResponseError> errors}) {
+    // if no ModelType fallback to default
+    if (request.modelType == null) {
+      return GraphQLResponse<T>(data: data as T, errors: errors);
+    }
+
+    if (request.decodePath == null) {
+      throw ApiException('No decodePath found',
+          recoverySuggestion: 'Include decodePath when creating a request');
+    }
+
+    Map<String, dynamic> dataJson = json.decode(data);
```

Comment:
A little tidbit about Dart and something to watch out for. There is an implicit cast happening here, which can throw in certain situations, if you're not careful. `json.decode` returns a `dynamic` which, in Dart, means it could be any type, even `null`. This statement here is the same as writing:

```dart
dynamic decoded = json.decode(data);
Map<String, dynamic> data = decoded as Map<String, dynamic>;
```

So, you can see the problem. `decoded` can be *any* JSON value (null, String, Map, etc.) and if it's not a Map<String, dynamic>, force casting with `as` will throw. There is an analyzer flag in dart which warns you about this because it can really bite you, but we don't have it on currently in the repo. I would recommend turning it on every once in a while developing because it will help you write safer, more robust code. Although, as you'll see, it's very strict so it's probably best to just use it sparingly and when you're curious. To turn it on, just add the following to an `analysis_options.yaml` file in either the project root or package root:

```yaml
analyzer:
  strong-mode:
    implicit-dynamic: false
    implicit-casts: false
```

With that turned on, you'll be *gently reminded* about this haha

<img width="1066" alt="Screen Shot 2021-07-30 at 8 11 10 AM" src="https://user-images.githubusercontent.com/24740863/127673989-35482b63-804d-491b-9629-1e83951f7ded.png">

Reply:
Very neat. I took it for a test drive and it lead me to create another null check. Thanks for the tip!